dimod/sampleset.py

```diff
@@ -1391,6 +1391,81 @@ def slice(self, *slice_args, **kwargs):
         return type(self)(record, self.variables, copy.deepcopy(self.info),
                           self.vartype)
 
+    def append_data_vectors(self, vectors, inplace=True):
+        """Return the :class:`SampleSet` with additional fields in
+        :attr:`SampleSet.record`.
+
+        Args:
+            vectors (dict):
+                Each key is a new field name and each value is a list of 
+                per-sample data to be appended to :attr:`SampleSet.record`. 
+                Per-sample data should be scalar or numpy arrays (lists 
+                and tuples will be converted to numpy arrays).
+
+            inplace (bool, optional, default=True):
```

Comment:
I would _not_ support this parameter, like in [.append_variables](https://github.com/dwavesystems/dimod/blob/4941d2301aca0dc2c61b586605d0d3c69e2a7320/dimod/sampleset.py#L1123). The sampleset is a fairly lightweight wrapper around the record array and IMO remaking the record should also just remake the entire object.

Reply:
This would also allow the user to pass in the vectors as kwargs. Which might be a nicer api. E.g. `new = sampleset.append_vectors(test=[1,2,3], another=[3, 4, 5])`.

I personally don't have much of an opinion between passing in dict or separate kwargs, but thought I would raise it as an option. Keeping the dict format is more future proof (easy to add unrelated kwargs later).

Reply:
Personally I like the idea of separate kwargs and agree that it's a nicer api. It doesn't seem very likely that we would need to add any unrelated kwargs in the future

Reply:
On that note (or notes), I wanted to suggest we rename this and/or add an update method, e.g. `update_vectors`, with a `dict.update`-like semantics.

`update_vectors` would always be inplace, and it would accept vectors as kwargs.

We can have `append_vectors` in parallel, and here I would leave the `inplace`, to be consistent with other `SampleSet` methods. Performance argument doesn't hold (if it's roughly the same perf), immutability by default if conceptually lovely, but that's not how SampleSet/BQM are currently designed. OTOH, `append_variables` doesn't have inplace either. :thinking:

Reply:
I think that because we cannot update the structure of the record in-place, we need to follow that behavior with the sampleset. Another approach would be to make `append_variables` and `append_vectors` functions rather than methods. 

Further, in the current implementation, it is not really safe to replace the `.record` object altogether (I can work up a failing example if you like).

Reply:
I think maybe having `append_variables` and `append_vectors` as functions might be more intuitive than having them as methods that don’t update in-place. 

On another note, I’m not sure if we should change `append_data_vectors` to `append_vectors`. We also have another method `SampleSet.data_vectors` which returns the data in the record.

Reply:
I am totally fine with making them functions (deprecation warning for the existing method of course).

Reply:
I guess it boils down to whether we want SampleSet to be a thin wrapper around numpy's recarray, or a (feature) rich container. I could easily argue for both, e.g. (1) savvy users with performance in mind would probably prefer thin wrapper over numpy; (2) users caring about API simplicity would probably appreciate rich feature set more than performance. The art, I guess, is in re-conciliating those two worlds.

Anyway, @arcondello, you're closer to dimod/SampleSet, so take my suggestions lightly. :)

Reply:
It seems to me that everyone here is fine with having `append_data_vectors` as a thin wrapper around recarray, so I’m going to go ahead and make it (and `append_variables`) a function that will return a new SampleSet. 

@radomir I would also be ok with having an additional update method that is always in place.